### DISCUSSION TITLE
Flake - unstable for nixvim and release for the rest

### DISCUSSION THREAD

**Discussion by ScOut3R** [nix-community/nixvim#1957] [Q&A] [answered]
Hi Nixvim community,

I'm new to NixOS and its ecosystem and found myself in trouble setting up Nixvim using the unstable branch of nixpkgs while keeping the release branch for the system and Home Manager.

When building the system I get the error `error: attribute 'cmp-ai' missing`, which indicates that Nixvim is looking at the latest stable nixpkgs which doesn't include the `cmp-ai` package yet.

I'm certain I'm doing something wrong in my flake. May I ask for your help?

```
{ description = "Flake";

  # This is the standard format for flake.nix. `inputs` are the dependencies of the flake,
  # and `outputs` function will return all the build results of the flake.
  # Each item in `inputs` will be passed as a parameter to the `outputs` function after being pulled and built.
  inputs = {
    # There are many ways to reference flake inputs. The most widely used is github:owner/name/reference,
    # which represents the GitHub repository URL + branch/commit-id/tag.

    # Official NixOS package source, using nixos-unstable branch here
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-24.05";
    nixpkgs-unstable.url = "github:NixOS/nixpkgs/nixos-unstable";

    [...]

    nixvim = {
      url = "github:nix-community/nixvim";
      inputs.nixpkgs.follows = "nixpkgs-unstable";
    };

    # home-manager, used for managing user configuration
    home-manager = {
      url = "github:nix-community/home-manager/release-24.05";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };

  # `outputs` are all the build result of the flake.
  # A flake can have many use cases and different types of outputs.
  # parameters in `outputs` are defined in `inputs` and can be referenced by their names.
  # However, `self` is an exception, This special parameter points to the `outputs` itself (self-reference)
  # The `@` syntax here is used to alias the attribute set of the inputs's parameter, making it convenient to use inside the function.
  outputs = { nixpkgs, nixpkgs-unstable, home-manager, ... }@inputs:
    {
      nixosConfigurations."carbon" = nixpkgs.lib.nixosSystem rec {
        system = "x86_64-linux";
        specialArgs = {
          inputs = inputs;
        };

        modules = [
          [...]

          home-manager.nixosModules.home-manager
          {
            home-manager.useGlobalPkgs = true;
            home-manager.useUserPackages = true;

            home-manager.users.[user] = import ./hosts/carbon/home.nix;

            # Optionally, use home-manager.extraSpecialArgs to pass arguments to home.nix
            home-manager.extraSpecialArgs = {
              inputs = inputs;
              nixpkgs-unstable = import nixpkgs-unstable { inherit system; };
            };
          }
        ];
      };
    };
}
```

**Comment by arunoruto** [comment]
Funny enough, I was thinking about this exact issue recently and was about to post the same question!
I would love to know if this is possible since I am eager to try the `lz-n` plugin!

**Comment by MattSturgeon** [comment, marked as answer]
`inputs.nixpkgs.follows` only has a meaningful effect when using nixvim's standalone build.

Your system is being build using `nixpkgs.lib.nixosSystem`, so the `pkgs` passed to modules will be derived from `nixpkgs` **not** `nixpkgs-unstable`.

There's currently no way to configure nixvim's modules to use a different `pkgs` to the host module's (see #1784), so the only way to achieve a nixvim built using `nixos-unstable` on a host system using `nixos-24.05` is to use a [standalone build](https://nix-community.github.io/nixvim/modules/standalone.html) of nixvim ([see also](https://nix-community.github.io/nixvim/#standalone-usage)).

**Reply by arunoruto** [reply]
Thanks a lot for the explanation! I just moved my nixvim config from home-manager to a standalone config using the flake template 😁 
I just had to delete the `programs.nixvim` to make it compatible with the standalone version and I can even compile the config a bit faster compared to home-manager.

**Reply by ScOut3R** [reply]
Thank you very much for your response @MattSturgeon!

It all makes sense and I learnt something fundamental about NixOS and flakes.
